sql/src/main/java/io/crate/metadata/functions/params/Param.java

```diff
@@ -0,0 +1,256 @@
+
+    private final SortedSet<DataType> validTypes;
+    private final ThreadLocal<FuncArg> boundType;
+    private final ThreadLocal<Deque<DataType>> multiBoundStack;
+
+    @Nullable
+    private final Param innerType;
+
+    private Param(DataType... validTypes) {
+        this(Collections.emptyList(), validTypes, null);
+    }
+
+    private Param(Collection<DataType> validTypes, DataType[] validTypes2, @Nullable Param innerType) {
+        this.boundType = new ThreadLocal<>();
+        this.validTypes = new TreeSet<>((o1, o2) -> {
+            if (o1.precedes(o2)) {
+                return -1;
+            } else if (o2.precedes(o1)) {
+                return 1;
+            }
+            return 0;
+        });
+        this.validTypes.addAll(validTypes);
+        this.validTypes.addAll(Arrays.asList(validTypes2));
+        this.multiBoundStack = ThreadLocal.withInitial(ArrayDeque::new);
+        this.innerType = innerType;
+    }
+
+    /**
+     * Creates a new Type with one or multiple valid {@link DataType}s.
+     * @param validTypes The valid {@link DataType}s.
+     * @return A new Type parameter.
+     */
+    public static Param of(DataType... validTypes) {
+        return new Param(validTypes);
+    }
+
+    /**
+     * Creates a new Type from a Collection and a list of {@link DataType}s.
+     * @param validTypes The valid {@link DataType} as a Collection.
+     * @param validTypes2 The valid {@link DataType}.
+     * @return A new Type parameter.
+     */
+    public static Param of(Collection<DataType> validTypes, DataType... validTypes2) {
+        return new Param(validTypes, validTypes2, null);
+    }
+
+    /**
+     * Matches the inner Type of this Type.
+     * @param innerType A Type to match against any inner type of a Type.
+     * @return A new Type with an inner Type.
+     */
+    public Param withInnerType(Param innerType) {
+        return new Param(this.validTypes, new DataType[]{}, innerType);
+    }
+
+    /**
+     * Returns the valid {@link DataType}s of this Type.
+     * @return A collection of valid types.
+     */
+    Collection<DataType> getValidTypes() {
+        return validTypes;
+    }
+
+    DataType getBoundType(boolean multiBind) {
+        final DataType boundType;
+        if (multiBind) {
+            boundType = this.multiBoundStack.get().removeFirst();
+        } else {
+            boundType = this.boundType.get().valueType();
+        }
+        Preconditions.checkState(boundType != null,
+            "Type not bound when it should have been.");
+        if (this.innerType != null) {
+            if (boundType instanceof CollectionType) {
+                DataType innerType = this.innerType.getBoundType(multiBind);
+                if (innerType != null) {
+                    return ((CollectionType) boundType).newInstance(innerType);
+                }
+            }
+        }
+        return boundType;
+    }
+
+    void bind(FuncArg funcArg, boolean multiBind) {
```

Comment:
This could also use some explanation - or maybe split it up a bit?

Reply:
I think `bindParam` and `rebindParam` would make this more clear (and might even help the JIT to inline - though that probably doesn't matter here)